### ISSUE TITLE
Virtual keyboard not supported on Debian

### ISSUE THREAD

**Issue by SamuelLongchamps** [SamuelLongchamps/OpenBoard#2]
The OS keyboard on which the current version (1.3 rc2) relies on for Linux is OnBoard, which is still in Debian testing. The default keyboard being Antler-based Caribou on Debian 8 and handled with the various focus events internally when activated through gnome's Accessibility feature makes it impractical to use Caribou for such feature (especially since it pops-out in a layout aimed at tablets, taking the bottom-half of the screen).

The goal would be to use a Qt implementation rather than rely on OS's specific keyboards for better control and homogeneousity among software platforms. The main features would be:
- Draggable
- Easily resizable
- Adapted to the activity (e.g. numpad standalone for math-based activities)

The goal is not to develop one from scratch, it would be wise to use an open third-party that we can use as a static library and expand as needed.

There is an option for a keyboard in the preference dialog, but we cannot expect the user to enable it if no OS keyboard is available. This keyboard is limited (not resizable for instance), which is why this task is to use a feature-complete third party otherwise (as the custom one has been reported for removal on the upstream project).

Quick fix is to check if onboard is installed upon startup of software and enable the custom keyboard if not. This is still viable if we choose to go with OS keyboards all the way considering that the next release of Debian will have onboard on its repo (current in testing).

**Comment by watsaig** [comment]
Well, I hadn't thought of compatibility with platforms that don't (easily) offer Onboard... Admittedly it would be better to check if it is installed on startup, rather than list it as a dependency in the .deb.
It would be very easy to hide the entry in the preferences dialog if there is no system OSK, for example.

By the way, we aren't necessarily bent on removing the in-built keyboard, but there are a couple reasons why we were thinking of it.

First, the OSs' OSKs are usually much better than any homebrew alternatives. There is merit to having a consistent UI, but in my opinion it isn't worth it to try and make it absolutely identical on all platforms. Especially when interfacing with the OS's native components such as an OSK or -- more commonly -- dialog boxes, it is usually OK (or even better) to use the native solution.

Second, it was starting to seem like fixing the built-in OSK would take a _lot_ of work. While it would be quite easy to have a simple keyboard with pre-defined layouts, this one doesn't do that. It queries the OS to fetch its current keyboard layouts, and uses those. So there is a different implementation for each platform. In this case it was misbehaving both on Linux and OS X, so it was deemed easier to extend support for system OSKs to Windows and OS X, and eventually offer only that, rather than fix

So anyway, it would take some work to get the built-in OSK working properly again, but if it's going to be fixed we can definitely leave it in :)
Unless you know of any good libraries to replace it?

**Comment by SamuelLongchamps** [comment]
I agree with all the reasons why the system OSK is a good choice to go. As I see it, choosing an external OSK or a Qt built one from an open source third party is as good if they are feature equivalent. A custom solution will have a hard time keeping up with other entirely dedicated solutions from third parties. The nice thing with OnBoard is the possibility to provide custom layouts as SVGs to implement the "adapted to the activity" feature suggested in the issue. A Qt-built one might be more complex, but will be available to all platforms which is not necessarily for the sake of homogeneity in term of UI consistency, but rather of feature consistency. Having to support different implementations of OSK for each platform is a complexity that should not be understated, especially on a longer term. Moreover, in the wake of tablets, system OSKs are becoming more like Caribou (taking half the screen), and this layout is far from optimal  for a smart board.

More locally and for shorter term, our problem is more in the event of unavailability of OnBoard on certain  GNU/Linux distributions like Debian stable. The software should be tolerant and force an alternative if OnBoard is not found. OnBoard is great at what it does, but until it hits Debian stable it should be listed as a suggested package and the backup implementation should be activated automatically if OnBoard fails to load/isn't installed. Teachers shouldn't be expected to enable the backup solution manually in the preferences dialog like it currently is.

I do not know of any Qt library for an OSK which would meet the current requirements, but I didn't take the time yet to search adequately. I know you referenced one in https://github.com/DIP-SEM/OpenBoard/issues/2 which has nice features like copy & paste. It would be interesting to experiment with it.
